### ISSUE TITLE
NNPA BERT Models Produce NaN Outputs for Zero-Filled Inputs When seq_len ≥ 64

### ISSUE THREAD

**Issue by KanakaPathivada** [onnx/onnx-mlir#3347]
**Summary**
When running BERT models compiled with NNPA using zDLC (ONNXMLIR backend) or PyTorch (torch nnpa), zero-filled inputs with sequence length **≥ 64** consistently produce **NaN outputs** and **ZDNN softmax runtime failures**.
CPU-compiled versions of the same model behave normally.

This appears to be an NNPA runtime or zDNN softmax kernel issue triggered specifically by zero-valued inputs at this sequence length boundary.

**Environment**
1. zDLC Image: `icr.io/ibmz/zdlc:5.0.0`
2. Model: bert-base-uncased (HuggingFace)
3. Backends tested:
- ONNXMLIR Triton backend (NNPA)
- PyTorch Triton backend (torch nnpa)
- CPU versions of both for comparison
4. Triton Inference Server: Triton IS with ONNXMLIR and PyTorch backends

**Issue Description**
1. BERT compiled for CPU → Works for all sequence lengths
2. BERT compiled for NNPA (ONNXMLIR / torch nnpa) →
- Works for seq_len ≤ 63
- Fails with NaN outputs for seq_len ≥ 64
- Logs show ZDNN softmax errors or range violations

**Observed Behavior Matrix**
Seq Length | CPU Model | NNPA ONNXMLIR | NNPA torch nnpa
-- | -- | -- | --
≤ 63 | ✔ Works | ✔ Works | ✔ Works
≥ 64 | ✔ Works | ❌ NaN outputs | ❌ zdnn_softmax failure

**Errors Observed**
ONNXMLIR Triton Backend (NNPA):
```
WARN: ZDNN_ELEMENT_RANGE_VIOLATION: Range violation on tensor data
```
PyTorch Backend (NNPA):
```
zdnn_softmax failed with rc 131073
```
Client Output Example (NNPA, seq_len=64)
```
"data": [NaN, NaN, NaN, ...]
```

**zDLC Compilation Commands:**
NNPA:
```
export ZDLC_IMAGE="icr.io/ibmz/zdlc:5.0.0" 
export ZDLC_MODEL_DIR=`pwd`
export ZDLC_MODEL_NAME="bert-base-uncased" 
podman run --rm -v ${ZDLC_MODEL_DIR}:/workdir:z ${ZDLC_IMAGE} --EmitLib --O3 --march=arch15 --mtriple=s390x-ibm-loz --maccel=NNPA --onnx-op-stats=JSON ${ZDLC_MODEL_NAME}.onnx
[2/6] Wed Dec  3 14:56:24 2025 (3s) Compiling and Optimizing MLIR Module
{
  "func.func" : 1,
  "func.return" : 1,
  "onnx.Add" : 14,
  "onnx.Cast" : 1,
  "onnx.Concat" : 132,
  "onnx.Constant" : 65,
  "onnx.Dim" : 205,
  "onnx.Gather" : 3,
  "onnx.LayerNormalization" : 26,
  "onnx.Mul" : 1,
  "onnx.NoValue" : 1,
  "onnx.Reshape" : 132,
  "onnx.Slice" : 1,
  "onnx.Sub" : 1,
  "onnx.Transpose" : 48,
  "onnx.Unsqueeze" : 2,
  "zhigh.Add" : 24,
  "zhigh.Gelu" : 13,
  "zhigh.MatMul" : 98,
  "zhigh.Softmax" : 12,
  "zhigh.Stick" : 86,
  "zhigh.StickifiedConstant" : 148,
  "zhigh.Unstick" : 98
}
[3/6] Wed Dec  3 14:56:28 2025 (7s) Translating MLIR Module to LLVM and Generating LLVM Optimized Bitcode
[4/6] Wed Dec  3 14:57:08 2025 (47s) Generating Object from LLVM Bitcode
[5/6] Wed Dec  3 14:57:26 2025 (65s) Linking and Generating the Output Shared Library
[6/6] Wed Dec  3 14:57:27 2025 (66s) Compilation completed 
```

CPU:
```
export ZDLC_IMAGE="icr.io/ibmz/zdlc:5.0.0" 
export ZDLC_MODEL_DIR=`pwd`
export ZDLC_MODEL_NAME="bert-base-uncased" 
podman run --rm -v ${ZDLC_MODEL_DIR}:/workdir:z ${ZDLC_IMAGE} --EmitLib --O3 --march=arch15 --mtriple=s390x-ibm-loz --onnx-op-stats=JSON ${ZDLC_MODEL_NAME}.onnx
[1/6] Wed Dec  3 19:04:03 2025 (0s) Importing ONNX Model to MLIR Module from "bert-base-uncased.onnx"
[2/6] Wed Dec  3 19:04:04 2025 (1s) Compiling and Optimizing MLIR Module
{
  "func.func" : 1,
  "func.return" : 1,
  "onnx.Add" : 112,
  "onnx.Cast" : 1,
  "onnx.Concat" : 48,
  "onnx.Constant" : 295,
  "onnx.Dim" : 97,
  "onnx.Gather" : 3,
  "onnx.Gelu" : 13,
  "onnx.LayerNormalization" : 26,
  "onnx.MatMul" : 98,
  "onnx.Mul" : 1,
  "onnx.Reshape" : 48,
  "onnx.Slice" : 1,
  "onnx.Softmax" : 12,
  "onnx.Sub" : 1,
  "onnx.Transpose" : 48,
  "onnx.Unsqueeze" : 2
}
[3/6] Wed Dec  3 19:04:08 2025 (5s) Translating MLIR Module to LLVM and Generating LLVM Optimized Bitcode
[4/6] Wed Dec  3 19:05:45 2025 (102s) Generating Object from LLVM Bitcode
[5/6] Wed Dec  3 19:06:51 2025 (168s) Linking and Generating the Output Shared Library
[6/6] Wed Dec  3 19:06:54 2025 (171s) Compilation completed 
```
`NOTE:` Disabling saturation (-nnpa-disable-saturation=true) does not resolve the issue.

**Reproduction Steps**

1. Compile bert-base-uncased for NNPA using zDLC.
2. Deploy the compiled model in Triton (either ONNXMLIR or PyTorch backend with NNPA).
3. Run the following inference with zero-filled input and seq_len = 64:
```
curl -X POST http://localhost:8000/v2/models/bert_model_nnpa/infer \
  -H "Content-Type: application/json" \
  -d '{
    "inputs": [
      { "name": "input_ids", "shape": [1, 64], "datatype": "INT64", "data": [[0...]] },
      { "name": "attention_mask", "shape": [1, 64], "datatype": "INT64", "data": [[0...]] },
      { "name": "token_type_ids", "shape": [1, 64], "datatype": "INT64", "data": [[0...]] }
    ],
    "outputs": [{ "name": "logits" }]
  }'
```
4. Observe NaN outputs or softmax execution errors.

**Expected Behavior**
- NNPA-compiled models should produce valid logits for any supported sequence length (including 64+).
- Behavior should match CPU-compiled BERT.
- No ZDNN softmax range violations or NaN outputs.

**Comment by pradghos** [comment]
FYI @Sunny-Anand

**Comment by Sunny-Anand** [comment]
@tungld can you pls take a look at this.

**Comment by tungld** [comment]
Are all "input_ids", "attention_mask" and "token_type_ids" zero-filled? Are they practical inputs, meaning, from what sentence the inputs were generated?

**Comment by tungld** [comment]
In bert models, when attention_mask contains only zero values, it means all values in the input_ids are padding values. So when doing masking, the output is a tensor of negative infinity values. Then, that tensor is the input of softmax. Softmax of negative infinity values will give us `0/0` which is invalid or NaN.

**Comment by KanakaPathivada** [comment]
Thank you for the clarification.

However, I would like to highlight one concern:

Even though an all-zero `attention_mask` technically represents a fully padded input, the **CPU execution path still produces valid (non-NaN) outputs** in this scenario. This behavior is important because Triton features such as **model warmup** and the **perf_analyzer’s** `--input-data zero` option rely on zero-filled inputs to validate and benchmark models.

With the current NNPA behavior—where zero-filled sequences of length ≥ 64 lead to NaN values or execution errors - these Triton features fail for NNPA-optimized models, even though they work correctly on CPU.

Because of this discrepancy, it may still be necessary to handle this case more gracefully on the NNPA path, or at least document and surface it clearly as a limitation, since users commonly rely on zero-input warmup and perf testing in production deployments.

**Comment by tungld** [comment]
As you observed, both onnx-mlir and torch_nnpa returned NaN because they used the same nnpa instruction and it's not onnx-mlir specific issue. The difference between CPU and NNPA is the way they handle infinity and NaN. On CPU, it can return infinity and you may not see NaN, but it does not mean the result is mathematically correct. On NNPA, NaN and infinity are fused into a single NaN as stated in this paper https://ieeexplore.ieee.org/document/8877411.

If you still want the same result as CPU in this case, the best way is to request a change in NNPA instruction to handle infinity similarly to CPU.

My recommendation is to use one-filled attention-mask instead of zero-filled attention-mask. Since I am not sure what is the correct result mathematically when attention-mask is zero, as I mentioned we will see `softmax(negative infinity)` and what is its correct result?

**Comment by KanakaPathivada** [comment]
Thank you for the explanation.

I understand that this behavior is inherent to NNPA and not specific to onnx-mlir, and that CPU and NNPA differ in how they handle infinity and NaN. That makes sense from an architectural perspective.

However, I want to highlight a practical limitation: in Triton, features such as **model warmup** and **perf_analyzer** with `--input-data zero`, users do not have an option to provide a one-filled attention mask. These workflows always generate zero-filled inputs by design.

As a result, NNPA-optimized models fail (NaN or execution errors) in standard validation and benchmarking flows, while the same models succeed on CPU. Even if the mathematical result of `softmax(-inf)` is undefined, this difference in behavior makes NNPA models incompatible with commonly used Triton features.

Given this, it would be helpful to either:
- Clearly document this as a known NNPA limitation, or
- Consider whether NNPA could handle this case more gracefully to align with existing Triton tooling expectations.

We will proceed with documenting this limitation on the Triton side, but wanted to confirm and track the behavior here.

**Comment by AlexandreEichenberger** [comment]
> However, I want to highlight a practical limitation: in Triton, features such as model warmup and perf_analyzer with --input-data zero, users do not have an option to provide a one-filled attention mask. These workflows always generate zero-filled inputs by design.

Maybe Triton should use models where 0 are legal... namely here add an "inverse attention mask" and add a "not" in the model before first use of the attention mask.  Or perform such negation in software prior to sending it to a backend.

Are there more actions to do before closing this issue? @tungld @KanakaPathivada
